vault/resource_azure_secret_backend_role.go

```diff
@@ -0,0 +1,210 @@
+				ForceNew:    true,
+				Description: "Human-friendly description of the mount for the backend.",
+			},
+			"azure_roles": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"role_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+
+						"role_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"scope": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"application_object_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Application Object ID for an existing service principal that will be used instead of creating dynamic service principals.",
+			},
+			"ttl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Human-friendly description of the mount for the backend.",
+			},
+			"max_ttl": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Human-friendly description of the mount for the backend.",
+			},
+		},
+	}
+}
+
+func azureSecretBackendRoleUpdateFields(d *schema.ResourceData, data map[string]interface{}) {
+
+	if v, ok := d.GetOk("azure_roles"); ok {
+		rawAzureList := v.(*schema.Set).List()
+
+		// Vaults API requires we send trhe policy as an escaped string
+		// So we marshall and then change into a string
+		jsonAzureList, _ := json.Marshal(rawAzureList)
+		jsonAzureListString := string(jsonAzureList)
+
+		log.Printf("[DEBUG] Azure RoleSet turned to escaped JSON: %s", jsonAzureListString)
+		data["azure_roles"] = string(jsonAzureListString)
+	}
+
+	if v, ok := d.GetOk("application_object_id"); ok {
+		data["application_object_id"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		data["ttl"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("max_ttl"); ok {
+		data["max_ttl"] = v.(string)
+	}
+
+}
+
+func azureSecretBackendRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+
+	path := azureSecretRoleResourcePath(backend, role)
+
+	data := map[string]interface{}{}
+	azureSecretBackendRoleUpdateFields(d, data)
+
+	log.Printf("[DEBUG] Writing role %q to Azure Secret backend", path)
+	d.SetId(path)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		d.SetId("")
+		return fmt.Errorf("Error writing Azure Secret role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Wrote role %q to Azure Secret backend", path)
+
+	return azureSecretBackendRoleRead(d, meta)
+}
+
+func azureSecretBackendRoleUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	data := map[string]interface{}{}
+	azureSecretBackendRoleUpdateFields(d, data)
+
+	log.Printf("[DEBUG] Updating role %q in Azure Secret backend", path)
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
```

Comment:
This is unimportant and the code looks great as-is. I wanted to share a tip here - sometimes with stuff like this I instead do:
```
if _, err := client.Logical().Write(path, data); err != nil {
    ...
}
```
The benefit is that the err then is only available inside the if statement. It's a slightly more defensive way to code because it doesn't leave the err available for somebody else to accidentally misuse lower down - like for them to accidentally use it in a different goroutine (which would cause a race), or to accidentally interpret an err from here as an err from there. 

Not at all a necessity, just thought I would share in case you felt like adopting this somewhere down the line. This code still looks very good and this is more cake frosting stuff.

Reply:
Yep, makes a lot of sense, happy to fix! 👍